From 46e8b8f2594a5596e6f9e22a83a2910d39793991 Mon Sep 17 00:00:00 2001
From: Vince Chang <vince_chang@aspeedtech.com>
Date: Wed, 28 May 2025 15:55:16 -0600
Subject: [PATCH] Add multi node support

Add node option to customize snoop object and dbus.

Tested:

```
root@ast2700-default:~# /usr/bin/snoopd -b 1 -d /dev/aspeed-lpc-snoop1 -n 1 &
root@ast2700-default:~# busctl tree xyz.openbmc_project.State.Boot.Raw1
`- /xyz
  `- /xyz/openbmc_project
    `- /xyz/openbmc_project/state
      `- /xyz/openbmc_project/state/boot
        `- /xyz/openbmc_project/state/boot/raw1
root@ast2700-default:~# snooper 1
node1_recv: 0x7e
node1_recv: 0xb0
node1_recv: 0x7e
node1_recv: 0xb1
node1_recv: 0xaf
```

Upstream-Status: Submitted [https://gerrit.openbmc.org/c/openbmc/phosphor-host-postd/+/80597]

Change-Id: I926ea01a542ab527f438a99875aeab75ef4f89d8
Signed-off-by: Vince Chang <vince_chang@aspeedtech.com>
---
 7seg.cpp                    |  5 ++++-
 example.cpp                 | 13 ++++++++++---
 lpcsnoop/snoop.hpp          |  4 ++--
 lpcsnoop/snoop_listen.hpp   | 25 +++++++++++++++++--------
 main.cpp                    | 27 +++++++++++++++++++++++----
 test/post_reporter_test.cpp | 10 +++++-----
 6 files changed, 61 insertions(+), 23 deletions(-)

diff --git a/7seg.cpp b/7seg.cpp
index a9247d2..c50c17b 100644
--- a/7seg.cpp
+++ b/7seg.cpp
@@ -49,6 +49,8 @@ static void DisplayDbusValue(FILE* f, postcode_t postcodes)
  */
 int main(int argc, const char* argv[])
 {
+    std::string node = "0";
+
     if (argc != 2 || !fs::exists(argv[1]))
     {
         std::fprintf(stderr, "usage: %s <device_node>\n", argv[0]);
@@ -60,7 +62,8 @@ int main(int argc, const char* argv[])
 
     auto ListenBus = sdbusplus::bus::new_default();
     std::unique_ptr<lpcsnoop::SnoopListen> snoop =
-        std::make_unique<lpcsnoop::SnoopListen>(ListenBus, DisplayDbusValue, f);
+        std::make_unique<lpcsnoop::SnoopListen>(ListenBus, DisplayDbusValue,
+                                                node, f);
 
     signal(SIGINT, [](int signum) {
         if (signum == SIGINT)
diff --git a/example.cpp b/example.cpp
index 95cc7a4..eab5136 100644
--- a/example.cpp
+++ b/example.cpp
@@ -21,6 +21,8 @@
 #include <iostream>
 #include <memory>
 
+static std::string node = "0";
+
 /* Example PostCode handler which simply prints them */
 static void printPostcode(FILE*, postcode_t postcode)
 {
@@ -32,7 +34,7 @@ static void printPostcode(FILE*, postcode_t postcode)
     {
         hexCode += std::format("{:02x}", byte);
     }
-    std::printf("recv: 0x%s\n", hexCode.c_str());
+    std::printf("node%s_recv: 0x%s\n", node.c_str(), hexCode.c_str());
 }
 
 /*
@@ -64,10 +66,15 @@ static void printPostcode(FILE*, postcode_t postcode)
  * This application simply creates an object that registers for incoming value
  * updates for the POST code dbus object.
  */
-int main()
+int main(int argc, char* argv[])
 {
+    if (argc > 1)
+    {
+        node = argv[1];
+    }
+
     auto ListenBus = sdbusplus::bus::new_default();
-    lpcsnoop::SnoopListen snoop(ListenBus, printPostcode);
+    lpcsnoop::SnoopListen snoop(ListenBus, printPostcode, node);
 
     while (true)
     {
diff --git a/lpcsnoop/snoop.hpp b/lpcsnoop/snoop.hpp
index 2d43251..a530e48 100644
--- a/lpcsnoop/snoop.hpp
+++ b/lpcsnoop/snoop.hpp
@@ -5,9 +5,9 @@
 #include <xyz/openbmc_project/State/Boot/Raw/server.hpp>
 
 /* The LPC snoop on port 80h is mapped to this dbus path. */
-constexpr char snoopObject[] = "/xyz/openbmc_project/state/boot/raw0";
+static std::string snoopObject = "/xyz/openbmc_project/state/boot/raw";
 /* The LPC snoop on port 80h is mapped to this dbus service. */
-constexpr char snoopDbus[] = "xyz.openbmc_project.State.Boot.Raw";
+static std::string snoopDbus = "xyz.openbmc_project.State.Boot.Raw";
 
 template <typename... T>
 using ServerObject = typename sdbusplus::server::object_t<T...>;
diff --git a/lpcsnoop/snoop_listen.hpp b/lpcsnoop/snoop_listen.hpp
index 241ee4f..fb04150 100644
--- a/lpcsnoop/snoop_listen.hpp
+++ b/lpcsnoop/snoop_listen.hpp
@@ -25,10 +25,15 @@ namespace lpcsnoop
 using std::get;
 
 /* Returns matching string for what signal to listen on Dbus */
-static const std::string GetMatchRule()
+static const std::string GetMatchRule(const std::string& node)
 {
     using namespace sdbusplus::bus::match::rules;
 
+    if (!node.empty())
+    {
+        snoopObject += node;
+    }
+
     return type::signal() + interface("org.freedesktop.DBus.Properties") +
            member("PropertiesChanged") + path(snoopObject);
 }
@@ -39,18 +44,22 @@ class SnoopListen
     using postcode_handler_t = std::function<void(FILE*, postcode_t)>;
 
   public:
-    SnoopListen(sdbusplus::bus_t& busIn, sd_bus_message_handler_t handler) :
-        signal(busIn, GetMatchRule().c_str(), handler, this)
+    SnoopListen(sdbusplus::bus_t& busIn, sd_bus_message_handler_t handler,
+                const std::string& node) :
+        signal(busIn, GetMatchRule(node).c_str(), handler, this)
     {}
 
-    SnoopListen(sdbusplus::bus_t& busIn, message_handler_t handler) :
-        signal(busIn, GetMatchRule(), handler)
+    SnoopListen(sdbusplus::bus_t& busIn, message_handler_t handler,
+                const std::string& node) :
+        signal(busIn, GetMatchRule(node), handler)
     {}
 
     SnoopListen(sdbusplus::bus_t& busIn, postcode_handler_t handler,
-                FILE* f = nullptr) :
-        SnoopListen(busIn, std::bind(defaultMessageHandler, handler, f,
-                                     std::placeholders::_1))
+                const std::string& node, FILE* f = nullptr) :
+        SnoopListen(
+            busIn,
+            std::bind(defaultMessageHandler, handler, f, std::placeholders::_1),
+            node)
     {}
 
     SnoopListen() = delete; // no default constructor
diff --git a/main.cpp b/main.cpp
index d4f4063..cbf2468 100644
--- a/main.cpp
+++ b/main.cpp
@@ -264,6 +264,7 @@ int main(int argc, char* argv[])
 {
     int postFd = -1;
     unsigned int rateLimit = 0;
+    static std::string node = "0";
 
     int opt;
 
@@ -275,6 +276,7 @@ int main(int argc, char* argv[])
         {"host", optional_argument, nullptr, 'h'},
 #else
         {"device", optional_argument, nullptr, 'd'},
+        {"node", optional_argument, nullptr, 'n'},
         {"rate-limit", optional_argument, nullptr, 'r'},
         {"bytes",  required_argument, nullptr, 'b'},
 #endif
@@ -287,7 +289,7 @@ int main(int argc, char* argv[])
 #ifdef ENABLE_IPMI_SNOOP
         "h:"
 #else
-        "d:r:b:"
+        "d:n:r:b:"
 #endif
         "v";
 
@@ -338,6 +340,12 @@ int main(int argc, char* argv[])
                     return -1;
                 }
                 break;
+            case 'n':
+            {
+                node = optarg;
+                fprintf(stderr, "Setting node to %s.\n", optarg);
+                break;
+            }
             case 'r':
             {
                 int argVal = -1;
@@ -384,12 +392,23 @@ int main(int argc, char* argv[])
 
     bool deferSignals = true;
 
+    // Add the node to the snoop object and dbus service names.
+    if (node == "0")
+    {
+        snoopObject += node;
+    }
+    else
+    {
+        snoopObject += node;
+        snoopDbus += node;
+    }
+
     // Add systemd object manager.
-    sdbusplus::server::manager_t snoopdManager(bus, snoopObject);
+    sdbusplus::server::manager_t snoopdManager(bus, snoopObject.c_str());
 
-    PostReporter reporter(bus, snoopObject, deferSignals);
+    PostReporter reporter(bus, snoopObject.c_str(), deferSignals);
     reporter.emit_object_added();
-    bus.request_name(snoopDbus);
+    bus.request_name(snoopDbus.c_str());
 
     // Create sdevent and add IO source
     try
diff --git a/test/post_reporter_test.cpp b/test/post_reporter_test.cpp
index 4a4f695..e94b018 100644
--- a/test/post_reporter_test.cpp
+++ b/test/post_reporter_test.cpp
@@ -34,7 +34,7 @@ TEST_F(PostReporterTest, EmitsObjectsOnExpectedDbusPath)
                 sd_bus_emit_object_added(IsNull(), StrEq(snoopObject)))
         .WillOnce(Return(0));
 
-    PostReporter testReporter(bus, snoopObject, true);
+    PostReporter testReporter(bus, snoopObject.c_str(), true);
     testReporter.emit_object_added();
 }
 
@@ -50,18 +50,18 @@ TEST_F(PostReporterTest, AddsObjectWithExpectedName)
                                          StrEq(snoopDbus), _, _))
         .WillOnce(slotcb);
 
-    PostReporter testReporter(bus, snoopObject, true);
+    PostReporter testReporter(bus, snoopObject.c_str(), true);
 }
 
 TEST_F(PostReporterTest, ValueReadsDefaultToEmpty)
 {
-    PostReporter testReporter(bus, snoopObject, true);
+    PostReporter testReporter(bus, snoopObject.c_str(), true);
     EXPECT_TRUE(std::get<0>(testReporter.value()).empty());
 }
 
 TEST_F(PostReporterTest, SetValueToPositiveValueWorks)
 {
-    PostReporter testReporter(bus, snoopObject, true);
+    PostReporter testReporter(bus, snoopObject.c_str(), true);
     primary_post_code_t primaryCode = {122, 126, 127};
     secondary_post_code_t secondaryCode = {123, 124, 125};
     testReporter.value(std::make_tuple(primaryCode, secondaryCode));
@@ -71,7 +71,7 @@ TEST_F(PostReporterTest, SetValueToPositiveValueWorks)
 
 TEST_F(PostReporterTest, SetValueMultipleTimesWorks)
 {
-    PostReporter testReporter(bus, snoopObject, true);
+    PostReporter testReporter(bus, snoopObject.c_str(), true);
     primary_post_code_t primaryCode = {20, 21, 0, 123};
     secondary_post_code_t secondaryCode = {10, 40, 0, 245, 56};
     testReporter.value(std::make_tuple(primaryCode, secondaryCode));
-- 
2.43.0

